frontend/packages/console-shared/src/hooks/perspective-utils.ts

```diff
@@ -0,0 +1,161 @@
+          acc[ex.properties.id] = true;
+          return acc;
+        },
+        {},
+      );
+    } else {
+      const perspectives: Perspective[] = JSON.parse(window.SERVER_FLAGS.perspectives);
+      obj = perspectiveExtensions.reduce((acc, perspectiveExtension) => {
+        const perspective = perspectives?.find((p) => p.id === perspectiveExtension.properties.id);
+
+        if (perspective?.visibility?.state === PerspectiveVisibilityState.Disabled) {
+          acc[perspectiveExtension.properties.id] = false;
+        } else {
+          acc[perspectiveExtension.properties.id] = true;
+        }
+        return acc;
+      }, {});
+    }
+    return obj;
+  });
+  const adminPerspective = perspectiveExtensions.find(
+    (perspectiveEx) => perspectiveEx.properties.id === 'admin',
+  );
+
+  const handleResults = React.useCallback(
+    (newResult: string, state: boolean) => {
+      setResults((oldResults: Record<string, boolean>) => {
+        return {
+          ...oldResults,
+          [newResult]: state,
+        };
+      });
+    },
+    [setResults],
+  );
+  React.useEffect(() => {
+    if (window.SERVER_FLAGS.perspectives) {
+      const perspectives: Perspective[] = JSON.parse(window.SERVER_FLAGS.perspectives);
+      perspectiveExtensions.forEach((perspectiveExtension) => {
+        const perspective = perspectives?.find((p) => p.id === perspectiveExtension.properties.id);
+
+        if (
+          !perspective ||
+          !perspective.visibility ||
+          perspective.visibility.state === PerspectiveVisibilityState.Enabled
+        ) {
+          handleResults(perspectiveExtension.properties.id, true);
+        } else if (
+          perspective?.visibility?.state === PerspectiveVisibilityState.AccessReview &&
+          perspective?.visibility?.accessReview &&
+          Object.keys(perspective?.visibility?.accessReview)?.length > 0
+        ) {
+          hasReviewAccess(perspective?.visibility?.accessReview)
+            .then((res) => {
+              if (res) {
+                handleResults(perspectiveExtension.properties.id, true);
+              } else {
+                handleResults(perspectiveExtension.properties.id, false);
+              }
+            })
+            .catch((e) => {
+              handleResults(perspectiveExtension.properties.id, true);
+              // eslint-disable-next-line no-console
+              console.warn('Could not check access', e);
+            });
+        }
+      });
+    }
+  }, [perspectiveExtensions, handleResults]);
+  const perspectives = React.useMemo(() => {
+    const filteredExtensions = perspectiveExtensions.filter((e) => results[e.properties.id]);
+
+    return filteredExtensions.length === 0 &&
+      Object.keys(results).length === perspectiveExtensions.length
+      ? [adminPerspective]
+      : filteredExtensions;
+  }, [perspectiveExtensions, results, adminPerspective]);
```

Comment:
You can inline the adminPerspective here.

Do we really need `filteredExtensions.length === 0` here? Doesn't this mean we see admin perspective until the permission checks didn't resolved?

```suggestion
  const perspectives = React.useMemo(() => {
    const filteredExtensions = perspectiveExtensions.filter((e) => results[e.properties.id]);

    return filteredExtensions.length === 0 &&
      Object.keys(results).length === perspectiveExtensions.length
      ? perspectiveExtensions.filter((p) => p.properties.id === 'admin')
      : filteredExtensions;
  }, [perspectiveExtensions, results, adminPerspective]);
```